server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,301 @@
+        );
+      }
+      catch (IOException e) {
+        throw new RE(e, "Failed to retrieve results from cache for query ID [%s]", query.getId());
+      }
+    });
+
+    return Sequences.map(cachedSequence, pullFromCacheFunction);
+  }
+
+  private ResultLevelCachePopulator createResultLevelCachePopulator(
+      String cacheKeyStr,
+      String resultSetId
+  )
+  {
+    if (resultSetId != null && populateResultCache) {
+      ResultLevelCachePopulator resultLevelCachePopulator = new ResultLevelCachePopulator(
+          cache,
+          objectMapper,
+          ResultLevelCacheUtil.computeResultLevelCacheKey(cacheKeyStr),
+          cacheConfig,
+          true
+      );
+      try {
+        //   Save the resultSetId and its length
+        resultLevelCachePopulator.cacheObjectStream.write(ByteBuffer.allocate(Integer.BYTES)
+                                                                    .putInt(resultSetId.length())
+                                                                    .array());
+        resultLevelCachePopulator.cacheObjectStream.write(StringUtils.toUtf8(resultSetId));
+      }
+      catch (IOException ioe) {
+        log.error("Failed to write cached values for query %s", query.getId());
+        return null;
+      }
+      return resultLevelCachePopulator;
+    } else {
+      return null;
+    }
+  }
+
+  public class ResultLevelCachePopulator
+  {
+    private final Cache cache;
+    private final ObjectMapper mapper;
+    private final Cache.NamedKey key;
+    private final CacheConfig cacheConfig;
+    private final ByteArrayOutputStream cacheObjectStream = new ByteArrayOutputStream();
+
+    public boolean isShouldPopulate()
+    {
+      return shouldPopulate;
+    }
+
+    private boolean shouldPopulate;
+
+    private ResultLevelCachePopulator(
+        Cache cache,
+        ObjectMapper mapper,
+        Cache.NamedKey key,
+        CacheConfig cacheConfig,
+        boolean shouldPopulate
+    )
+    {
+      this.cache = cache;
+      this.mapper = mapper;
+      this.key = key;
+      this.cacheConfig = cacheConfig;
+      this.shouldPopulate = shouldPopulate;
+    }
+
+    private void cacheResultEntry(
+        ResultLevelCachePopulator resultLevelCachePopulator,
+        T resultEntry,
+        Function<T, Object> cacheFn
+    )
+    {
+
+      int cacheLimit = cacheConfig.getResultLevelCacheLimit();
+      if (cacheLimit > 0 && resultLevelCachePopulator.cacheObjectStream.size() > cacheLimit) {
+        shouldPopulate = false;
+        return;
+      }
+      try (JsonGenerator gen = mapper.getFactory().createGenerator(resultLevelCachePopulator.cacheObjectStream)) {
+        gen.writeObject(cacheFn.apply(resultEntry));
+      }
+      catch (IOException ex) {
+        log.error("Failed to retrieve entry to be cached. Result Level caching will not be performed!");
+        shouldPopulate = false;
+      }
+    }
+
+    public void populateResults()
+    {
+      ResultLevelCacheUtil.populate(
```

Comment:
it might be cleaner to do the check whether we crossed limit at this point. I know its done externally but may be clearer to do that here.